compiler/rustc_middle/src/mir/syntax.rs

```diff
@@ -578,43 +577,13 @@ pub enum TerminatorKind<'tcx> {
     /// > The drop glue is executed if, among all statements executed within this `Body`, an assignment to
     /// > the place or one of its "parents" occurred more recently than a move out of it. This does not
     /// > consider indirect assignments.
-    Drop { place: Place<'tcx>, target: BasicBlock, unwind: Option<BasicBlock> },
-
-    /// Drops the place and assigns a new value to it.
-    ///
-    /// This first performs the exact same operation as the pre drop-elaboration `Drop` terminator;
-    /// it then additionally assigns the `value` to the `place` as if by an assignment statement.
-    /// This assignment occurs both in the unwind and the regular code paths. The semantics are best
-    /// explained by the elaboration:
-    ///
-    /// ```ignore (MIR)
-    /// BB0 {
-    ///   DropAndReplace(P <- V, goto BB1, unwind BB2)
-    /// }
-    /// ```
-    ///
-    /// becomes
-    ///
-    /// ```ignore (MIR)
-    /// BB0 {
-    ///   Drop(P, goto BB1, unwind BB2)
-    /// }
-    /// BB1 {
-    ///   // P is now uninitialized
-    ///   P <- V
-    /// }
-    /// BB2 {
-    ///   // P is now uninitialized -- its dtor panicked
-    ///   P <- V
-    /// }
-    /// ```
-    ///
-    /// Disallowed after drop elaboration.
-    DropAndReplace {
+    Drop {
         place: Place<'tcx>,
-        value: Operand<'tcx>,
         target: BasicBlock,
         unwind: Option<BasicBlock>,
+        /// This field is only used for better diagnostic and to check MIR drop invariants.
+        /// It has no effect on the semantics of the `Drop` terminator.
```

Comment:
Aren't Drop invariants are part of the semantics of the `Drop` terminator? If I understand correctly, incorrectly setting this flag to `true` could cause invalid mir to be accepted by the borrow checker (if we assume that there's a later assignment when there actually isn't).

Reply:
Mmm, good question.
> incorrectly setting this flag to true could cause invalid mir to be accepted by the borrow checker (if we assume that there's a later assignment when there actually isn't).

I would say that, by definition, changing the value of `is_replace` shouldn't change any property of a program.
If it does, it's because of implementation details (which are not relevant at this level).
We could even remove the field and check whether the next statement is an assignment to the same place everywhere we need the value of the flag.

The behavior in the borrowck is a result of the current implementation, not intrinsic in the definition of the `Drop` terminator. (e.g. my previous implementation didn't have this problem and the definition of `Drop` hasn't changed). In other words, the borrowck should check the access regardless of `is_replace`.


Actually, I would say that the semantics of `Drop` being dependent on `is_replace` would suggest that the removal of `DropAndReplace` was not very effective

Reply:
To avoid such issues and keep good diagnostics at the same time, you could change mir building to add some information to the spans that are used for drop terminators: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/struct.Span.html#method.mark_with_reason

Reply:
> We could even remove the field and check whether the next statement is an assignment to the same place everywhere we need the value of the flag.

That would be fragile to changes in MIR building I think.